Simplify and correct the Grep exercise (#2105)
exercism/problem-specifications: exercises/grep/description.md

# Description

Search files for lines matching a search string and return all matching lines.

> The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command searches for lines in files that match a regular expression.

Reviewer: @hkBst Did you dislike the use of "pattern" in the original description?

Author: I was not so happy that "pattern" was used ambiguously to refer both to "regular expression" and to "literal string", so I tried to clarify that while the real grep searches for regular expression( pattern)s, in this exercise you only need to implement literal string( pattern) search. The ambiguity of the word "pattern" made me try to avoid its use.

Reviewer: I think the ambiguity is fine, since below it's clearly stated that this exercise is a simplification and only cares about literal strings. If people are unfamiliar with grep they can look at the linked documentation.

My vote would be to go with `pattern` here.

Reviewer 2: And the fact that grep does match patterns, both with literal strings and regular expressions, it covers it (and reiterates what we find to be described in the first(?) line of the documentation.

Author: What exactly are you proposing? Do you just want to add the term "pattern"? Or do you want it to replace something else? And to what purpose?

The ambiguity about regular expression vs literal strings was perhaps my main motivation to take a look at improving the prose of this exercise. I am glad that you think this is now very clear, but also a bit afraid to lose this clarification.

Author: On further reflection I am happy to replace "regular expression" here with "pattern". Can you propose a change?

Reviewer: I don't feel strongly about this at all, I was mostly just curious why you preferred regex to pattern when I had the exact opposite reaction :) Usually when this happens I learn something interesting from asking the question.

I have a very weak preference towards:

> The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command searches for lines in files that match a pattern.

But having had this discussion I'm perfectly happy with the current state of the sentence.

Reviewer 3: This can be read as "files that match a regular expression" instead of "lines that match a regular expression"

Perhaps this (which hand-waves over grep processing its stdin, or handling multiple regexes):

> The grep command searches files for lines matching a regular expression.

I would also prefer "pattern" over "regular expression", mostly because `grep -F` exists to match fixed strings.

> The grep command searches files for lines matching a pattern.

Reviewer: Ah good catch on that one.

> The grep command searches files for lines matching a pattern.

This is now my favorite.

Author: Since fixed strings are a special case of regular expressions, I don't think we gain anything by being non-specific with "pattern". If anyone wonders what a pattern is and they click through, they find out that it is basically a regular expression, although grep supports two kinds of regular expressions and there is special support for easily matching on fixed strings. So, therefore, I wonder why not just say regular expression, when that is basically all grep does?

I do agree that this word reordering is a definite improvement.

Author: One more data point: I think the wikipedia page for regular expression is much more relevant here than the one for pattern.

RE: A regular expression is a sequence of [characters](https://en.wikipedia.org/wiki/Character_(computing)) that specifies a [search pattern](https://en.wikipedia.org/wiki/Pattern_matching) in [text](https://en.wikipedia.org/wiki/String_(computer_science)).
Pattern: A pattern is a regularity in the world, in human-made design, or in [abstract](https://en.wikipedia.org/wiki/Abstraction) ideas.

Reviewer: That works for me.

Your task is to implement a simplified `grep` command, which supports searching for literal strings.

The `grep` command takes three arguments:

1. The string to search for.
2. Zero or more flags for customizing the command's behavior.
3. One or more files to search in.

It then reads the contents of the specified files (in the order specified), finds the lines that contain the search string, and finally returns those lines (in the order in which they were found) prepended, when searching in multiple files, by the file name and a colon (':').

## Flags

The `grep` command should support the following flags:

> - `-n` Prepend the line number and a colon (':') to each line in the output. (Comes after possible file name.)

Reviewer: ```suggestion
- `-n` Prepend the line number and a colon (':') to each line in the output, placing the number after the filename (if present).
```

> - `-l` Instead of lines, only output the names of the files that contain at least one matching line.

Reviewer: ```suggestion
- `-l` Output only the names of the files that contain at least one matching line.
```

- `-i` Match using a case-insensitive comparison.
- `-v` Invert the program -- collect all lines that fail to match the pattern.

> - `-x` Search instead for lines matching the search string entirely, instead of merely containing it.

Reviewer: ```suggestion
- `-x` Search only for lines where the search string matches the entire line.
```

> # Instructions

Reviewer 4: This really is a description file. The heading is stripped out when imported into a track. It should not be changed to `instructions`.

Author: Done.

Search files for lines matching a search string and return all matching lines.

> The Unix [`grep`](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html) command searches for lines in files that match a regular expression. Your task is to implement a simplified `grep` function, which supports searching for literal strings.

Reviewer 4: Can you move the second sentence (`Your task`) to the next line?

As we're updating this file, it should now follow [the Exercism Markdown style](https://exercism.org/docs/building/markdown/markdown#layout).

Author: Done.

> The `grep` function takes three arguments:

Reviewer 4: It's a command. That in rust (perhaps?) you implement an equally named _function_ is an implementation detail. In various tracks, like JavaScript, you write a working program that has a `grep` command (not a function).

Author: Done.

1. The string to search for.

> 2. Zero or more flags to customize its behavior.

Reviewer 4: In English this `its` refers to `flags`, not `grep`. Repeating "the matching" resolves that. But if you say `matching`, then `pattern` is perhaps a better candidate because of pattern matching.

Author: Is it not true that since "flags" is plural it would need to be referred to with "their", and thus cannot be the target of "its"? Anyway, since it is potentially confusing, I've replaced "its" with "the command's".

Reviewer 4: Eh. Yeah you're right. I'm that case it would refer to the last singular noun which is not the command so your solution solves all of it. 🙏🏽

3. One or more files to search in.

It then reads the contents of the specified files (in the order specified),
finds the lines that contain the search string,

> and finally returns those lines (in the order in which they were found).

Reviewer 4: Can you make this a single line?

Author: Done.

## Flags

> The `grep` function should support the following flags:

Reviewer 4: ```suggestion
The `grep` command should support the following flags:
```

Author: Done.

> - `-n` Prepend the line number to each line in the output.

Reviewer 4: What happened to the "multiple files" note?

Author: I've now included this information in line 14 and this flag's description.
